R/data.table.R

```diff
@@ -936,6 +936,42 @@ replace_dot_alias = function(e) {
         else
           x
       }
+      # extract the same codes from setDT() and do_j_names() for #6702
+      simple_extract <- function(name, x, checkSingleChar = TRUE, envir = parent.frame()) {
+        k = eval(name[[2L]], envir, envir)
+        if (is.list(k)) {
+          origj = j = if (name[[1L]] == "$") {
+            as.character(name[[3L]])
+          } else {
+            eval(name[[3L]], envir, envir)
+          }
+          if (is.character(j)) {
+            if (checkSingleChar) {
+              if (length(j) != 1L) {
+                stopf("Cannot assign to an under-allocated recursively indexed list -- L[[i]][,:=] syntax is only valid when i is length 1, but its length is %d", length(j))
+              }
+              j2 = match(j, names(k))
+              if (is.na(j2)) {
```

Comment:
in both branches, we do the `j2 = match()` and `if (is.na())` checks -- move them outside `checkSingleChar`. Try and think through if that applies elsewhere also.